File: server/src/main/java/server/models/accountant/fund/Fund.java
package server.models.accountant.fund;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import server.models.Account;
import server.models.Document;

import java.time.LocalDateTime;

@Entity
@Table(name = "funds")
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Fund {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private Double balance;

    @Column(nullable = false)
    private String status = "ACTIVE";

    @Column(columnDefinition = "TEXT")
    private String purpose;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_account_id")
    private Account createdBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "updated_by_account_id")
    private Account updatedBy;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @OneToOne
    @JoinColumn(name = "document_id", unique = true)
    private Document document;

    @Column(name = "locked_until")
    private LocalDateTime lockedUntil;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: mobile/lib/models/fund.dart
class Fund {
  final int id;
  final String name;
  final double balance;
  final String status;
  final String? purpose;
  final String? createdByName;
  final String? updatedByName;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final DateTime? lockedUntil;

  Fund({
    required this.id,
    required this.name,
    required this.balance,
    required this.status,
    this.purpose,
    this.createdByName,
    this.updatedByName,
    required this.createdAt,
    this.updatedAt,
    this.lockedUntil,
  });

  factory Fund.fromJson(Map<String, dynamic> json) {
    return Fund(
      id: json['id'] is int
          ? json['id']
          : int.tryParse(json['id'].toString()) ?? 0,
      name: json['name'] ?? '',
      balance: (json['balance'] is int || json['balance'] is double)
          ? (json['balance'] ?? 0).toDouble()
          : double.tryParse(json['balance'].toString()) ?? 0.0,
      status: json['status'] ?? '',
      purpose: json['purpose'],
      createdByName: json['createdBy'] ?? json['createdByName'],
      updatedByName: json['updatedBy'] ?? json['updatedByName'],
      createdAt: DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'])
          : null,
      lockedUntil: json['lockedUntil'] != null
          ? DateTime.tryParse(json['lockedUntil'])
          : (json['lockDate'] != null
                ? DateTime.tryParse(json['lockDate'])
                : null),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "name": name,
      "balance": balance,
      "status": status,
      "purpose": purpose,
      "createdBy": createdByName,
      "updatedBy": updatedByName,
      "createdAt": createdAt.toIso8601String(),
      "updatedAt": updatedAt?.toIso8601String(),
      "lockedUntil": lockedUntil?.toIso8601String(),
    };
  }
}
